From 777ab89f70f80f7469206757f24c2948485c5e4e Mon Sep 17 00:00:00 2001
From: Josh Poimboeuf <jpoimboe@redhat.com>
Date: Thu, 4 Sep 2014 18:54:23 -0400
Subject: [kernel] ftrace: One more missing sync after fixup of function modification failure

Message-id: <cb2986e478e3032f966bc6618e8be784440842d2.1409856554.git.jpoimboe@redhat.com>
Patchwork-id: 91239
O-Subject: [RHEL7.1 PATCH 04/23] ftrace/x86: One more missing sync after fixup of function modification failure
Bugzilla: 1113829
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1113829

commit 12729f14d8357fb845d75155228b21e76360272d
Author: Petr Mladek <pmladek@suse.cz>
Date:   Mon Feb 24 17:12:20 2014 +0100

    ftrace/x86: One more missing sync after fixup of function modification failure

    If a failure occurs while modifying ftrace function, it bails out and will
    remove the tracepoints to be back to what the code originally was.

    There is missing the final sync run across the CPUs after the fix up is done
    and before the ftrace int3 handler flag is reset.

    Here's the description of the problem:

     CPU0    CPU1
     ----    ----
      remove_breakpoint();
      modifying_ftrace_code = 0;

        [still sees breakpoint]
        <takes trap>
        [sees modifying_ftrace_code as zero]
        [no breakpoint handler]
        [goto failed case]
        [trap exception - kernel breakpoint, no
         handler]
        BUG()

    Link: http://lkml.kernel.org/r/1393258342-29978-2-git-send-email-pmladek@suse.cz

    Fixes: 8a4d0a687a5 "ftrace: Use breakpoint method to update ftrace caller"
    Acked-by: Frederic Weisbecker <fweisbec@gmail.com>
    Acked-by: H. Peter Anvin <hpa@linux.intel.com>
    Signed-off-by: Petr Mladek <pmladek@suse.cz>
    Signed-off-by: Steven Rostedt <rostedt@goodmis.org>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/ftrace.c b/arch/x86/kernel/ftrace.c
index 6b566c8..69885e2 100644
--- a/arch/x86/kernel/ftrace.c
+++ b/arch/x86/kernel/ftrace.c
@@ -660,8 +660,8 @@ ftrace_modify_code(unsigned long ip, unsigned const char *old_code,
   ret = -EPERM;
   goto out;
  }
- run_sync();
  out:
+ run_sync();
  return ret;
 
  fail_update:
-- 
1.7.1